opentracing/propagator.py

```diff
@@ -0,0 +1,115 @@
+# Copyright (c) 2016 The OpenTracing Authors.
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+from __future__ import absolute_import
+
+from .span import Span
+
+class SpanPropagator(object):
+    """SpanPropagator encodes and decodes Spans between processes.
+
+    SpanPropagator is responsible (a) for encoding Span instances in a manner
+    suitable for propagation, and (b) for taking that encoded data and using it
+    to generate Span instances that are placed appropriately in the overarching
+    Trace. Typically the propagation will take place across an RPC boundary, but
+    message queues and other IPC mechanisms are also reasonable places to use
+    a SpanPropagator.
+
+    The encoded form of a propagated span is divided into two components:
+
+      1) The Tracer-internal state ("Tracer State") for the Span (for example,
+         in Dapper this would include a trace_id, a span_id, and a bitmask
+         representing the sampling status for the given trace)
+      2) Any trace attributes (per Span.set_trace_attribute)
+
+    The encoded data is separated in this way for a variety of reasons; the
+    most important is to give OpenTracing users a portable way to opt out of
+    Trace Attribute propagation entirely if they deem it a stability risk.
+
+    The propagate_span_as*() and join_trace_from_*() methods come in two
+    flavors: binary and text. The text format is better-suited to
+    pretty-printing and debugging, and the binary format is better-suited to
+    compact, high-performance encoding, decoding, and transmission.
```

Comment:
The last sentence is somewhat dubious, the key/value text format is needed not because of readability concerns, but because of the requirements of the underlying protocol, e.g. HTTP headers are key/value map. The actual `value` can still be a binary string encoded in base64. Not a huge deal, of course, since this will get rewritten via injectors later.